activation/handler_v1.go

```diff
@@ -244,25 +245,30 @@ func (h *HandlerV1) validateNonInitialAtx(
 		return err
 	}
 
-	nonce := atx.VRFNonce
-	if atx.NumUnits > previous.NumUnits && nonce == nil {
-		h.log.WithContext(ctx).With().Info("post size increased without new vrf Nonce, re-validating current nonce",
-			atx.ID(),
-			log.Stringer("smesher", atx.SmesherID),
-		)
-
-		// This is not expected to happen very often, so we query the database
-		// directly here without using the cache.
-		current, err := atxs.NonceByID(h.cdb, previous.ID())
-		if err != nil {
-			return fmt.Errorf("failed to get current nonce: %w", err)
+	needRecheck := atx.VRFNonce != nil || atx.NumUnits > previous.NumUnits
+	if atx.VRFNonce == nil {
+		// We load the previous ATX from blob and hence it might not contain a VRF nonce.
+		// In this case, we need to fetch the current nonce from the DB.
+		// The value 0 means that most likely it is not set and we need to query the DB.
+		if previous.VRFNonce == 0 {
+			current, err := atxs.NonceByID(h.cdb, previous.ID())
+			if err != nil {
+				return fmt.Errorf("failed to get current nonce from previous %s: %w", previous.ID(), err)
+			}
+			atx.VRFNonce = (*uint64)(&current)
+		} else {
+			atx.VRFNonce = (*uint64)(&previous.VRFNonce)
 		}
```

Comment:
I know this saves one DB query, but only in the case where the smesher updated its VRF nonce after they just updated it in the previous ATX. I feel this is an optimisation that will rarely apply and therefore can be left out:

```suggestion
		current, err := atxs.NonceByID(h.cdb, previous.ID())
		if err != nil {
			return fmt.Errorf("failed to get current nonce from previous %s: %w", previous.ID(), err)
		}
		atx.VRFNonce = (*uint64)(&current)
```

Reply:
This also saves a query for every second (after the initial one) ATX.

Reply:
Also note there may be a subtle issue with this construct
```go
			atx.VRFNonce = (*uint64)(&previous.VRFNonce)
```
This prevents the whole `previous` structure from being GC'd.
In this particular case `atx` itself is probably short-lived, so not really a problem, and the code is to be removed, but generally using pointers to members of other structs for fields which are pointers b/c they're optional should be avoided

Reply:
> generally using pointers to members of other structs for fields which are pointers b/c they're optional should be avoided

Good point, I updated the code to avoid that.

Reply:
> This also saves a query for every second (after the initial one) ATX.

`previous` comes from the DB right? So if the value for the `VRFNonce` is 0 querying again will just return 0 again won't it?

Reply:
The comment explains it:
```
// We load the previous ATX from blob and hence it might not contain a VRF nonce.
// In this case, we need to fetch the current nonce from the DB.
// The value 0 means that most likely it is not set and we need to query the DB.
```

The previous ATX is loaded from the blob* because the `numUnits` in the `atxs` table contains the _effective_ value. Here we need to compare the exact values (not _effective_ ones).

When it has a 0 value, it is highly probable that reading again from `atxs` would return a different value.

* unless it's a checkpointed ATX, when we fallback to fetching it from the `atxs` table.

Reply:
OK, so then the issue is that `previous` is of type `types.ActivationTx` and not `wire.ActivationTxV1`. Would the the ambiguity with effective Num Units and VRF nonce go away if we do not convert to a domain type?

We could decide to keep the blobs for the golden ATXs and only delete ATXs older than the new first epoch then no special handling of golden ATXs would be necessary in the ATX handler - but I would not be a fan of this decision.

Alternatively if `previous` is a golden ATX it would be `nil` and the nonce would be needed to be fetched via `atxs.VRFNonce` getting rid of the double meaning of 0. In the case of effective num units we would need to skip:

https://github.com/spacemeshos/go-spacemesh/blob/995d1be3de901050492eae8bf44a6587fc426980/activation/handler_v1.go#L188

and instead just take the numunits value of the current ATX which in the worst case gives a smesher more weight than they deserve in an epoch that (at the time the ATX became golden) would have already passed.

For weight concentration attacks it's worse if we underestimate a smeshers weight, not if we overestimate it. For rewards it doesn't matter: the weight in a previous epoch has no influence on current rewards and a golden ATX will always be a few epochs back in a checkpoint restore. Wdyt?